NPE in HTTP2 StreamResetTest.testClientResetConsumesQueuedData

**Jetty version(s)**
12

**Jetty Environment**
all

**Java version/vendor** `(use: java -version)`

**OS type/version**

**Description*

Whilst building a javadoc only change: https://jenkins.webtide.net/blue/organizations/jenkins/jetty.project/detail/PR-10527/5/pipeline/

```
[WARNING] Tests run: 5, Failures: 0, Errors: 0, Skipped: 1, Time elapsed: 15.19 s -- in org.eclipse.jetty.http2.tests.MaxConcurrentStreamsTest
2023-09-17 19:40:53.633:INFO :oejs.Server:ForkJoinPool-1-worker-4: Stopped oejs.Server@5a705a25{STOPPING}[12.0.2-SNAPSHOT,sto=0]
2023-09-17 19:40:53.633:INFO :oejs.AbstractConnector:ForkJoinPool-1-worker-4: Stopped ServerConnector@4173b1bb{h2c, (h2c)}{0.0.0.0:0}
2023-09-17 19:40:53.636:INFO :oejs.Server:ForkJoinPool-1-worker-4: jetty-12.0.2-SNAPSHOT; built: 2023-09-17T19:24:59.154Z; git: ad08b01f2e1a48a4dd46fc2826f5d9c3569edc9a; jvm 17.0.8+7
2023-09-17 19:40:53.638:INFO :oejs.AbstractConnector:ForkJoinPool-1-worker-4: Started ServerConnector@1fbf0f15{h2, (h2)}{0.0.0.0:38411}
2023-09-17 19:40:53.638:INFO :oejs.Server:ForkJoinPool-1-worker-4: Started oejs.Server@6d7bd9fc{STARTING}[12.0.2-SNAPSHOT,sto=0] @86209ms
2023-09-17 19:40:54.646:INFO :oejs.Server:ForkJoinPool-1-worker-4: Stopped oejs.Server@6d7bd9fc{STOPPING}[12.0.2-SNAPSHOT,sto=0]
2023-09-17 19:40:54.646:INFO :oejs.AbstractConnector:ForkJoinPool-1-worker-4: Stopped ServerConnector@1fbf0f15{h2, (h2)}{0.0.0.0:0}
2023-09-17 19:40:54.650:INFO :oejs.Server:ForkJoinPool-1-worker-4: jetty-12.0.2-SNAPSHOT; built: 2023-09-17T19:24:59.154Z; git: ad08b01f2e1a48a4dd46fc2826f5d9c3569edc9a; jvm 17.0.8+7
2023-09-17 19:40:54.653:INFO :oejs.AbstractConnector:ForkJoinPool-1-worker-4: Started ServerConnector@480bfcd6{h2, (h2)}{0.0.0.0:42721}
2023-09-17 19:40:54.653:INFO :oejs.Server:ForkJoinPool-1-worker-4: Started oejs.Server@5e2969e3{STARTING}[12.0.2-SNAPSHOT,sto=0] @87224ms
2023-09-17 19:40:57.160:INFO :oejs.Server:ForkJoinPool-1-worker-4: Stopped oejs.Server@5e2969e3{STOPPING}[12.0.2-SNAPSHOT,sto=0]
2023-09-17 19:40:57.160:INFO :oejs.AbstractConnector:ForkJoinPool-1-worker-4: Stopped ServerConnector@480bfcd6{h2, (h2)}{0.0.0.0:0}
2023-09-17 19:40:57.163:INFO :oejs.Server:ForkJoinPool-1-worker-4: jetty-12.0.2-SNAPSHOT; built: 2023-09-17T19:24:59.154Z; git: ad08b01f2e1a48a4dd46fc2826f5d9c3569edc9a; jvm 17.0.8+7
2023-09-17 19:40:57.164:INFO :oejs.AbstractConnector:ForkJoinPool-1-worker-4: Started ServerConnector@72bf65a8{h2, (h2)}{0.0.0.0:35889}
2023-09-17 19:40:57.164:INFO :oejs.Server:ForkJoinPool-1-worker-4: Started oejs.Server@182b5240{STARTING}[12.0.2-SNAPSHOT,sto=0] @89735ms
2023-09-17 19:40:59.171:INFO :oejs.Server:ForkJoinPool-1-worker-4: Stopped oejs.Server@182b5240{STOPPING}[12.0.2-SNAPSHOT,sto=0]
2023-09-17 19:40:59.171:INFO :oejs.AbstractConnector:ForkJoinPool-1-worker-4: Stopped ServerConnector@72bf65a8{h2, (h2)}{0.0.0.0:0}
2023-09-17 19:40:59.175:INFO :oejs.Server:ForkJoinPool-1-worker-4: jetty-12.0.2-SNAPSHOT; built: 2023-09-17T19:24:59.154Z; git: ad08b01f2e1a48a4dd46fc2826f5d9c3569edc9a; jvm 17.0.8+7
2023-09-17 19:40:59.177:INFO :oejs.AbstractConnector:ForkJoinPool-1-worker-4: Started ServerConnector@178e10f5{h2, (h2)}{0.0.0.0:45199}
2023-09-17 19:40:59.177:INFO :oejs.Server:ForkJoinPool-1-worker-4: Started oejs.Server@24c55980{STARTING}[12.0.2-SNAPSHOT,sto=0] @91748ms
2023-09-17 19:41:01.184:INFO :oejs.Server:ForkJoinPool-1-worker-4: Stopped oejs.Server@24c55980{STOPPING}[12.0.2-SNAPSHOT,sto=0]
2023-09-17 19:41:01.185:INFO :oejs.AbstractConnector:ForkJoinPool-1-worker-4: Stopped ServerConnector@178e10f5{h2, (h2)}{0.0.0.0:0}
2023-09-17 19:41:01.189:INFO :oejs.Server:ForkJoinPool-1-worker-4: jetty-12.0.2-SNAPSHOT; built: 2023-09-17T19:24:59.154Z; git: ad08b01f2e1a48a4dd46fc2826f5d9c3569edc9a; jvm 17.0.8+7
2023-09-17 19:41:01.190:INFO :oejs.AbstractConnector:ForkJoinPool-1-worker-4: Started ServerConnector@28ab56e5{h2, (h2)}{0.0.0.0:43899}
2023-09-17 19:41:01.190:INFO :oejs.Server:ForkJoinPool-1-worker-4: Started oejs.Server@3d92874{STARTING}[12.0.2-SNAPSHOT,sto=0] @93761ms
2023-09-17 19:41:02.198:INFO :oejs.Server:ForkJoinPool-1-worker-4: Stopped oejs.Server@3d92874{STOPPING}[12.0.2-SNAPSHOT,sto=0]
2023-09-17 19:41:02.199:INFO :oejs.AbstractConnector:ForkJoinPool-1-worker-4: Stopped ServerConnector@28ab56e5{h2, (h2)}{0.0.0.0:0}
2023-09-17 19:41:02.203:INFO :oejs.Server:ForkJoinPool-1-worker-4: jetty-12.0.2-SNAPSHOT; built: 2023-09-17T19:24:59.154Z; git: ad08b01f2e1a48a4dd46fc2826f5d9c3569edc9a; jvm 17.0.8+7
2023-09-17 19:41:02.205:INFO :oejs.AbstractConnector:ForkJoinPool-1-worker-4: Started ServerConnector@383f097b{h2, (h2)}{0.0.0.0:42439}
2023-09-17 19:41:02.205:INFO :oejs.Server:ForkJoinPool-1-worker-4: Started oejs.Server@2e512670{STARTING}[12.0.2-SNAPSHOT,sto=0] @94776ms
processing request /1
processing request /2
2023-09-17 19:41:03.713:WARN :oejuts.AdaptiveExecutionStrategy:server-4008: Task run failed
java.lang.NullPointerException: Cannot read field "_callback" because "this._request" is null
	at org.eclipse.jetty.server.internal.HttpChannelState.lambda$onIdleTimeout$1(HttpChannelState.java:402)
	at org.eclipse.jetty.util.thread.strategy.AdaptiveExecutionStrategy.runTask(AdaptiveExecutionStrategy.java:478)
	at org.eclipse.jetty.util.thread.strategy.AdaptiveExecutionStrategy.consumeTask(AdaptiveExecutionStrategy.java:441)
	at org.eclipse.jetty.util.thread.strategy.AdaptiveExecutionStrategy.tryProduce(AdaptiveExecutionStrategy.java:293)
	at org.eclipse.jetty.util.thread.strategy.AdaptiveExecutionStrategy.produce(AdaptiveExecutionStrategy.java:195)
	at org.eclipse.jetty.http2.HTTP2Connection.produce(HTTP2Connection.java:211)
	at org.eclipse.jetty.http2.HTTP2Connection.onFillable(HTTP2Connection.java:158)
	at org.eclipse.jetty.http2.HTTP2Connection$FillableCallback.succeeded(HTTP2Connection.java:449)
	at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:100)
	at org.eclipse.jetty.io.SelectableChannelEndPoint$1.run(SelectableChannelEndPoint.java:53)
	at org.eclipse.jetty.util.thread.strategy.AdaptiveExecutionStrategy.runTask(AdaptiveExecutionStrategy.java:478)
	at org.eclipse.jetty.util.thread.strategy.AdaptiveExecutionStrategy.consumeTask(AdaptiveExecutionStrategy.java:441)
	at org.eclipse.jetty.util.thread.strategy.AdaptiveExecutionStrategy.tryProduce(AdaptiveExecutionStrategy.java:293)
	at org.eclipse.jetty.util.thread.strategy.AdaptiveExecutionStrategy.run(AdaptiveExecutionStrategy.java:201)
	at org.eclipse.jetty.util.thread.ReservedThreadExecutor$ReservedThread.run(ReservedThreadExecutor.java:410)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:971)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.doRunJob(QueuedThreadPool.java:1196)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.run(QueuedThreadPool.java:1151)
	at java.base/java.lang.Thread.run(Thread.java:833)
2023-09-17 19:41:03.713:WARN :oejuts.AdaptiveExecutionStrategy:server-4005: Task run failed
java.lang.NullPointerException: Cannot read field "_callback" because "this._request" is null
	at org.eclipse.jetty.server.internal.HttpChannelState.lambda$onIdleTimeout$1(HttpChannelState.java:402)
	at org.eclipse.jetty.util.thread.strategy.AdaptiveExecutionStrategy.runTask(AdaptiveExecutionStrategy.java:478)
	at org.eclipse.jetty.util.thread.strategy.AdaptiveExecutionStrategy.consumeTask(AdaptiveExecutionStrategy.java:441)
	at org.eclipse.jetty.util.thread.strategy.AdaptiveExecutionStrategy.tryProduce(AdaptiveExecutionStrategy.java:293)
	at org.eclipse.jetty.util.thread.strategy.AdaptiveExecutionStrategy.produce(AdaptiveExecutionStrategy.java:195)
	at org.eclipse.jetty.http2.HTTP2Connection.produce(HTTP2Connection.java:211)
	at org.eclipse.jetty.http2.HTTP2Connection.onFillable(HTTP2Connection.java:158)
	at org.eclipse.jetty.http2.HTTP2Connection$FillableCallback.succeeded(HTTP2Connection.java:449)
	at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:100)
	at org.eclipse.jetty.io.SelectableChannelEndPoint$1.run(SelectableChannelEndPoint.java:53)
	at org.eclipse.jetty.util.thread.strategy.AdaptiveExecutionStrategy.runTask(AdaptiveExecutionStrategy.java:478)
	at org.eclipse.jetty.util.thread.strategy.AdaptiveExecutionStrategy.consumeTask(AdaptiveExecutionStrategy.java:441)
	at org.eclipse.jetty.util.thread.strategy.AdaptiveExecutionStrategy.tryProduce(AdaptiveExecutionStrategy.java:293)
	at org.eclipse.jetty.util.thread.strategy.AdaptiveExecutionStrategy.run(AdaptiveExecutionStrategy.java:201)
	at org.eclipse.jetty.util.thread.ReservedThreadExecutor$ReservedThread.run(ReservedThreadExecutor.java:410)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:971)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.doRunJob(QueuedThreadPool.java:1196)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.run(QueuedThreadPool.java:1151)
	at java.base/java.lang.Thread.run(Thread.java:833)
2023-09-17 19:41:03.947:INFO :oejs.Server:ForkJoinPool-1-worker-4: Stopped oejs.Server@2e512670{STOPPING}[12.0.2-SNAPSHOT,sto=0]
2023-09-17 19:41:03.947:INFO :oejs.AbstractConnector:ForkJoinPool-1-worker-4: Stopped ServerConnector@383f097b{h2, (h2)}{0.0.0.0:0}
[INFO] Tests run: 7, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 27.93 s -- in org.eclipse.jetty.http2.tests.IdleTimeoutTest
[INFO] 
[INFO] Results:
[INFO] 
[ERROR] Failures: 
[ERROR]   StreamResetTest.testClientResetConsumesQueuedData:443 expected: <0> but was: <32768>
```

--- a/jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/internal/HttpChannelState.java
+++ b/jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/internal/HttpChannelState.java
@@ -1403,6 +1403,8 @@ public class HttpChannelState implements HttpChannel, Components
 
         MetaData.Response lockedPrepareResponse(HttpChannelState httpChannel, boolean last)
         {
+            assert _request._lock.isHeldByCurrentThread();
+
             // Assume 200 unless told otherwise.
             if (_status == 0)
                 _status = HttpStatus.OK_200;
@@ -1537,6 +1539,9 @@ public class HttpChannelState implements HttpChannel, Components
             try (AutoLock ignored = _request._lock.lock())
             {
                 httpChannelState = _request._httpChannelState;
+                if (httpChannelState == null)
+                    return; // channel already completed
+
                 stream = httpChannelState._stream;
                 request = _request;
 
@@ -1639,6 +1644,7 @@ public class HttpChannelState implements HttpChannel, Components
         @Override
         MetaData.Response lockedPrepareResponse(HttpChannelState httpChannelState, boolean last)
         {
+            assert httpChannelState._request._lock.isHeldByCurrentThread();
             MetaData.Response httpFields = super.lockedPrepareResponse(httpChannelState, last);
             httpChannelState._response._status = _status;
             HttpFields.Mutable originalResponseFields = httpChannelState._responseHeaders.getMutableHttpFields();

--- a/jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/internal/HttpChannelState.java
+++ b/jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/internal/HttpChannelState.java
@@ -396,10 +396,11 @@ public class HttpChannelState implements HttpChannel, Components
                 }
 
                 // otherwise, if there is no failure listener, then we can fail the callback directly without a double lock
-                if (_onFailure == null && _request != null)
+                ChannelRequest request = _request;
+                if (_onFailure == null && request != null)
                 {
                     _failure = Content.Chunk.from(t, true);
-                    return () -> _request._callback.failed(t);
+                    return () -> request._callback.failed(t);
                 }
             }
         }

--- a/jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/internal/HttpChannelState.java
+++ b/jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/internal/HttpChannelState.java
@@ -1538,15 +1538,12 @@ public class HttpChannelState implements HttpChannel, Components
             ErrorResponse errorResponse = null;
             try (AutoLock ignored = _request._lock.lock())
             {
+                if (lockedCompleteCallback())
+                    return;
                 httpChannelState = _request._httpChannelState;
-                if (httpChannelState == null)
-                    return; // channel already completed
-
                 stream = httpChannelState._stream;
                 request = _request;
 
-                if (lockedCompleteCallback())
-                    return;
                 assert httpChannelState._callbackFailure == null;
 
                 httpChannelState._callbackFailure = failure;

